cmd/object-api-utils.go

```diff
@@ -565,29 +571,82 @@ func NewGetObjectReader(rs *HTTPRangeSpec, oi ObjectInfo, pcfn CheckCopyPrecondi
 		// encrypted bytes. The header parameter is used to
 		// provide encryption parameters.
 		fn = func(inputReader io.Reader, h http.Header, pcfn CheckCopyPreconditionFn, cFns ...func()) (r *GetObjectReader, err error) {
-			copySource := h.Get(crypto.SSECopyAlgorithm) != ""
-
 			cFns = append(cleanUpFns, cFns...)
-			// Attach decrypter on inputReader
-			var decReader io.Reader
-			decReader, err = DecryptBlocksRequestR(inputReader, h,
-				off, length, seqNumber, partStart, oi, copySource)
-			if err != nil {
-				// Call the cleanup funcs
+			runCleanUp := func() {
 				for i := len(cFns) - 1; i >= 0; i-- {
 					cFns[i]()
 				}
+			}
+
+			// First, we extract and decrypt the object encryption key.
+			// This process is slightly different for SSE-C and SSE-S3
+			// since for the former we need the client key (in the HTTP
+			// headers) and for the later we need to talk to the KMS.
+			var objectEncrypionKey crypto.ObjectKey
+			switch {
+			case crypto.S3.IsEncrypted(oi.UserDefined):
+				objectEncrypionKey, err = crypto.S3.UnsealObjectKey(GlobalKMS, oi.UserDefined, oi.Bucket, oi.Name)
+				if err != nil {
+					runCleanUp()
+					return nil, err
+				}
+			case crypto.SSEC.IsEncrypted(oi.UserDefined):
+				// In case of SSE-C the client may either perform a regular GET
+				// or a COPY. If the client provides the SSE-C copy headers we
+				// have to use them instead of the regular SSE-C headers.
+				// For more details see: https://docs.aws.amazon.com/AmazonS3/latest/dev/ServerSideEncryptionCustomerKeys.html
+				if crypto.SSECopy.IsRequested(h) {
+					objectEncrypionKey, err = crypto.SSECopy.UnsealObjectKey(h, oi.UserDefined, oi.Bucket, oi.Name)
+				} else {
+					objectEncrypionKey, err = crypto.SSEC.UnsealObjectKey(h, oi.UserDefined, oi.Bucket, oi.Name)
+				}
+				if err != nil {
+					runCleanUp()
+					return nil, err
+				}
+			default:
+				runCleanUp()
+				return nil, errObjectTampered
+			}
+
+			// Finally, we need to decrypt the ETag. Unfortunately, S3 ETag handling is complicated.
+			// In general, the ETag is encrypted (to prevent leaking information about the  object
+			// content through it). According to the AWS S3, the ETag (sent to the client) is the
+			// object MD5 in case of SSE-S3 but not in case of SSE-C (or SSE-KMS).
+			//
+			// However, we have to always preserve the MD5 of the object content since a client could
+			// copy an SSE-C-encrypted object (ETag != MD5(content)) to an SSE-S3-encrypted object
+			// (ETag == MD5(content)). ¯\_(ツ)_/¯
+			//
+			// Therefore, we return the decrypted ETag to the client whenever the encryption "method"
+			// is SSE-S3. Otherwise, we return the (a part) of the encrypted ETag to the client.
```

Comment:
I don't understand what `we return the (a part) of the encrypted` means.

Reply:
The encrypted etag consists of `<header><encrypted-plaintext><authentication-tag>`. So the encrypted etag is always lager then 16 bytes. However, the client expects a 16byte etag.

Tried to clarify in comment.